### PR TITLE
feat!: Add comment to PR

### PR DESCRIPTION
**Pull request by akash1810**
## What does this change?
In #20 a job summary was added with Riff-Raff information.

This change takes this a little further, adding a comment to the pull request with links to deploy the project in Riff-Raff. Only one comment will be made. As new builds are created, the comment is updated.

This change adds two new inputs:
  - (required) `githubToken` is a GitHub token with permissions to comment on a pull request.
  - (optional) `commentingStage` which stage to deploy to. Default of `CODE`.

## How to test
Use the action!

- https://github.com/guardian/service-catalogue/pull/487 would be a typical usage, only setting `githubToken`.
- https://github.com/guardian/cdk-playground/pull/399 sets `commentingStage` to `PROD` as the app only has one stage.

## Have we considered potential risks?
This should be released as a breaking change as a new required input `githubToken` is added.

## Images
image

<details><summary>Draft release notes</summary>
<p>

# v3.0.0
- Add a pull request comment with quick links to deploy with Riff-Raff

## BREAKING CHANGES
A new required input `githubToken` has been added. This needs to be scoped with `pull-requests: write` permission.

### Before (v2)
```yaml
jobs:
  CI:
    timeout-minutes: 15
    runs-on: ubuntu-latest
    steps:
      - uses: guardian/actions-riff-raff@v2
        with:
          projectName: playground::my-application
          configPath: riff-raff.yaml
          contentDirectories: |
            static-site-assets:
              - static-site/dist
            api:
              - target/application.jar
```

### After (v3)
```yaml
jobs:
  CI:
    timeout-minutes: 15
    runs-on: ubuntu-latest
    permissions: 
      pull-requests: write # <- NEW!
    steps:
      - uses: guardian/actions-riff-raff@v3 # Note the version change
        with:
          githubToken: ${{ secrets.GITHUB_TOKEN }} # <- NEW!
          projectName: playground::my-application
          configPath: riff-raff.yaml
          contentDirectories: |
            static-site-assets:
              - static-site/dist
            api:
              - target/application.jar
```

</p>
</details>

### CODE REVIEW

File: dist/index.js
Comment (NovemberTang) [inline dist/index.js:1]: did you intend to commit this file?

Comment (akash1810) [reply dist/index.js:1]: Yep. From the [instructions in the README](https://github.com/guardian/actions-riff-raff?tab=readme-ov-file#local-development):

> Edit the Typescript as usual and **remember to build** (`npm run build`) before committing to ensure `index.js` is updated.

This file needs to be committed as its [what gets executed](https://github.com/guardian/actions-riff-raff/blob/79c1742e6316f566f2e71b1ec19710ca7c00bad4/action.yaml#L34) when the action is used.

File: README.md [hunk 1]
```diff
@@ -114,6 +118,16 @@ first build produced by this action will be 46.
...
 
+### `githubToken`
+*Required*
+
+A GitHub token scoped to allow pull request commenting.
+
+### `commentingStage`
+_Default: CODE_
+
+When commenting on a pull request, which stage should be used. Typically a pre-production stage.
```
Comment (akash1810) [inline README.md:129]: I'm not convinced this is the best variable name. Alternatives welcomed!

**Review by JuliaBrigitte — APPROVED** [review]

**Comment by rtyley** [comment]
> Only one comment will be made. As new builds are created, the comment is updated.

@akash1810 is this still intended to be the case? Unfortunately it doesn't seem to be working with the recent PRs I've seen. For instance, in this PR: https://github.com/guardian/ophan/pull/6106, there are **8** comments made by `actions-riff-raff`:

 * [Deploy build 23997 of ophan:dashboard-es7-arm to CODE](https://riffraff.gutools.co.uk/deployment/deployAgain?project=ophan%3Adashboard-es7-arm&build=23997&stage=CODE&updateStrategy=MostlyHarmless&action=deploy)
 * [Deploy build 23997 of ophan:the-slab to CODE](https://riffraff.gutools.co.uk/deployment/deployAgain?project=ophan%3Athe-slab&build=23997&stage=CODE&updateStrategy=MostlyHarmless&action=deploy)
 * [Deploy build 23997 of ophan:big-loader-es7-arm to CODE](https://riffraff.gutools.co.uk/deployment/deployAgain?project=ophan%3Abig-loader-es7-arm&build=23997&stage=CODE&updateStrategy=MostlyHarmless&action=deploy)
 * etc

It's quite noisy, and it would be great if all these could be in the same comment, rather than 8 separate comments!


image

**Comment by akash1810** [comment]
@rtyley the action is behaving as intended - a comment is added for each Riff-Raff project an artifact has been built for. In the case of Ophan, there are 8 unique projects being built, hence 8 comments.

Folding this into a single comment sounds great! Indeed it has [come up previously](https://github.com/guardian/actions-riff-raff/pull/95#issuecomment-1861152471).

Looking at https://github.com/guardian/riffraff-platform/blob/main/packages/common/src/access.ts, there are more repositories that produce more than 1 artifact, which could help prioritise implementing this.

Let's create an issue for this?

**Pull request merged**